File: server/src/main/com/broll/gainea/server/core/actions/required/SelectChoiceAction.java
package com.broll.gainea.server.core.actions.required;

import com.broll.gainea.net.NT_Action_SelectChoice;
import com.broll.gainea.net.NT_Reaction;
import com.broll.gainea.server.core.actions.AbstractActionHandler;
import com.broll.gainea.server.core.actions.ActionContext;
import com.broll.gainea.server.core.actions.RequiredActionContext;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.utils.PlayerUtils;
import com.broll.networklib.server.impl.LobbyPlayer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.stream.Collectors;

public class SelectChoiceAction extends AbstractActionHandler<NT_Action_SelectChoice, SelectChoiceAction.Context> {

    private final static Logger Log = LoggerFactory.getLogger(SelectChoiceAction.class);

    class Context extends ActionContext<NT_Action_SelectChoice> {
        Player selectingPlayer;
        int selectedOption;

        public Context(NT_Action_SelectChoice action) {
            super(action);
        }
    }

    public int selection(String message, List<String> choices) {
        return selection(game.getCurrentPlayer(), message, choices);
    }

    public int selection(Player player, String message, List<String> choices) {
        Log.debug("Select choice action");
        Context context = build(choices);
        actionHandlers.getReactionActions().requireAction(player, new RequiredActionContext<>(context, message));
        Log.trace("Wait for select choice reaction");
        processingBlock.waitFor(player);
        return context.selectedOption;
    }

    public int selectObject(String message, List<? extends Object> choices) {
        return selectObject(game.getCurrentPlayer(), message, choices);
    }

    public int selectObject(Player player, String message, List<? extends Object> choices) {
        assureNotEmpty(choices);
        Log.debug("Select object action");
        NT_Action_SelectChoice action = new NT_Action_SelectChoice();
        action.objectChoices = choices.toArray(new Object[0]);
        Context context = new Context(action);
        actionHandlers.getReactionActions().requireAction(player, new RequiredActionContext<>(context, message));
        Log.trace("Wait for select choice reaction");
        processingBlock.waitFor(player);
        return context.selectedOption;
    }

    public Location selectLocation(String message, List<? extends Location> choices) {
        return selectLocation(game.getCurrentPlayer(), message, choices);
    }

    public Location selectLocation(Player player, String message, List<? extends Location> choices) {
        assureNotEmpty(choices);
        Log.debug("Select location action");
        NT_Action_SelectChoice action = new NT_Action_SelectChoice();
        action.objectChoices = choices.stream().map(Location::getNumber).toArray();
        Context context = new Context(action);
        actionHandlers.getReactionActions().requireAction(player, new RequiredActionContext<>(context, message));
        Log.trace("Wait for select choice reaction");
        processingBlock.waitFor(player);
        return choices.get(context.selectedOption);
    }

    public Player selectOtherPlayer(Player player, String message) {
        return selectPlayer(player, PlayerUtils.getOtherPlayers(game, player).collect(Collectors.toList()), message);
    }

    public Player selectPlayer(Player player, List<Player> options, String message) {
        Log.debug("Select player action");
        Context context = build(options.stream().map(Player::getServerPlayer).map(LobbyPlayer::getName).collect(Collectors.toList()));
        context.selectingPlayer = player;
        actionHandlers.getReactionActions().requireAction(player, new RequiredActionContext<>(context, message));
        Log.trace("Wait for select choice reaction");
        processingBlock.waitFor(player);
        return options.get(context.selectedOption);
    }

    private void assureNotEmpty(List choices) {
        if (choices.isEmpty()) {
            throw new RuntimeException("Invalid select choice context: list of choices is empty");
        }
    }

    private Context build(List<String> choices) {
        assureNotEmpty(choices);
        NT_Action_SelectChoice action = new NT_Action_SelectChoice();
        action.choices = choices.toArray(new String[0]);
        return new Context(action);
    }

    @Override
    public void handleReaction(Context context, NT_Action_SelectChoice action, NT_Reaction reaction) {
        Log.trace("Handle select choice reaction");
        context.selectedOption = reaction.option;
        processingBlock.resume();
    }

}

File: server/src/main/com/broll/gainea/server/core/actions/required/SelectChoiceAction.kt
package com.broll.gainea.server.core.actions.required

import com.broll.gainea.misc.RandomUtils
import com.broll.gainea.net.NT_Action_SelectChoice
import com.broll.gainea.net.NT_Reaction
import com.broll.gainea.server.core.actions.AbstractActionHandler
import com.broll.gainea.server.core.actions.ActionContext
import com.broll.gainea.server.core.actions.RequiredActionContext
import com.broll.gainea.server.core.actions.required.SelectChoiceAction.Context
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.utils.getOtherPlayers
import org.slf4j.LoggerFactory

class SelectChoiceAction :
    AbstractActionHandler<NT_Action_SelectChoice, Context>() {
    inner class Context(
        action: NT_Action_SelectChoice,
        var selectedOption: Int = 0
    ) : ActionContext<NT_Action_SelectChoice>(action)

    fun selection(message: String, choices: List<String>): Int {
        return selection(game.currentPlayer, message, choices)
    }

    fun selection(player: Player, message: String, choices: List<String>): Int {
        Log.debug("Select choice action")
        val context = build(choices)
        actionHandlers.reactionActions.requireAction(
            player,
            RequiredActionContext(context, message)
        )
        Log.trace("Wait for select choice reaction")
        processingBlock.waitFor(player)
        if (processingBlock.playerLeft) {
            context.selectedOption = RandomUtils.randomIndex(choices)
        }
        return context.selectedOption
    }

    fun selectObject(message: String, choices: List<Any>): Int {
        return selectObject(game.currentPlayer, message, choices)
    }

    fun selectObject(player: Player, message: String, choices: List<Any>): Int {
        assureNotEmpty(choices)
        Log.debug("Select object action")
        val action = NT_Action_SelectChoice()
        action.objectChoices = choices.toTypedArray<Any>()
        val context = Context(action)
        actionHandlers.reactionActions.requireAction(
            player,
            RequiredActionContext(context, message)
        )
        Log.trace("Wait for select choice reaction")
        processingBlock.waitFor(player)
        if (processingBlock.playerLeft) {
            context.selectedOption = RandomUtils.randomIndex(choices)
        }
        return context.selectedOption
    }

    fun selectLocation(message: String, choices: List<Location>): Location {
        return selectLocation(game.currentPlayer, message, choices)
    }

    fun selectLocation(player: Player, message: String, choices: List<Location>): Location {
        assureNotEmpty(choices)
        Log.debug("Select location action")
        val action = NT_Action_SelectChoice()
        val distinctLocations = choices.distinct()
        action.objectChoices = distinctLocations.map { it.number }.toTypedArray()
        val context = Context(action)
        actionHandlers.reactionActions.requireAction(
            player,
            RequiredActionContext(context, message)
        )
        Log.trace("Wait for select choice reaction")
        processingBlock.waitFor(player)
        if (processingBlock.playerLeft) {
            context.selectedOption = RandomUtils.randomIndex(choices)
        }
        return distinctLocations[context.selectedOption]
    }

    fun selectOtherPlayer(player: Player, message: String): Player {
        return selectPlayer(player, game.getOtherPlayers(player), message)
    }

    fun selectPlayer(player: Player, options: List<Player>, message: String): Player {
        Log.debug("Select player action")
        val context = build(options.map { it.serverPlayer.name })
        actionHandlers.reactionActions.requireAction(
            player,
            RequiredActionContext(context, message)
        )
        Log.trace("Wait for select choice reaction")
        processingBlock.waitFor(player)
        if (processingBlock.playerLeft) {
            context.selectedOption = RandomUtils.randomIndex(options)
        }
        return options[context.selectedOption]
    }

    private fun assureNotEmpty(choices: List<*>) {
        if (choices.isEmpty()) {
            throw RuntimeException("Invalid select choice context: list of choices is empty")
        }
    }

    private fun build(choices: List<String>): Context {
        assureNotEmpty(choices)
        val action = NT_Action_SelectChoice()
        action.choices = choices.toTypedArray<String>()
        return Context(action)
    }

    override fun handleReaction(
        context: Context,
        action: NT_Action_SelectChoice,
        reaction: NT_Reaction
    ) {
        Log.trace("Handle select choice reaction")
        context.selectedOption = reaction.option
        processingBlock.resume()
    }
    
    companion object {
        private val Log = LoggerFactory.getLogger(SelectChoiceAction::class.java)
    }


}
